src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,268 @@
+
+        if (!lastPropertyAccessChain) return undefined;
+
+        const occurrences = getOccurrencesInExpression(lastPropertyAccessChain.expression, expression.left, checker);
+        return occurrences ? { lastPropertyAccessChain, occurrences, expression } : undefined;
+    }
+
+    /**
+     * Gets a list of property accesses that appear in matchTo and occur in sequence in expression.
+     */
+    function getOccurrencesInExpression(matchTo: Expression, expression: Expression, checker: TypeChecker): (PropertyAccessExpression | Identifier)[] | undefined {
+        const occurrences: (PropertyAccessExpression | Identifier)[] = [];
+        while (isBinaryExpression(expression) && expression.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken) {
+            const match = getMatchingSubexpression(matchTo, expression.right, checker);
+            if (!match) {
+                break;
+            }
+            occurrences.push(match);
+            matchTo = match;
+            expression = expression.left;
+        }
+        const finalMatch = getMatchingSubexpression(matchTo, expression, checker);
+        if (finalMatch) {
+            occurrences.push(finalMatch);
+        }
+        return occurrences.length > 0 ? occurrences: undefined;
+    }
+
+    /**
+     * Checks that expression is a syntactic subexpression of matchTo.
+     */
+    function getMatchingSubexpression(matchTo: Expression, expression: Expression, checker: TypeChecker): PropertyAccessExpression | Identifier | undefined {
+        if (isCallExpression(matchTo)) {
+            return !isCallExpression(expression) ? getMatchingSubexpression(matchTo.expression, expression, checker) : undefined;
+        }
+        else if ((isPropertyAccessExpression(expression) || isIdentifier(expression)) && (isPropertyAccessExpression(matchTo) || isIdentifier(matchTo))) {
+            return containsSyntacticSubchain(matchTo, expression, checker) ? expression : undefined;
+        }
+        return undefined;
+    }
+
+    /**
+     * Returns true if target is a syntactic subchain of source.
+     */
+    function containsSyntacticSubchain(source: Node, target: Node, checker: TypeChecker): boolean {
+        while (isPropertyAccessExpression(source) && !isSyntacticMatch(source, target)) {
+            source = source.expression;
+        }
+        while (isPropertyAccessExpression(source) && isPropertyAccessExpression(target)) {
+            if (!isSyntacticMatch(source, target)) return false;
+            source = source.expression;
+            target = target.expression;
+        }
+        const fullMatch = isSyntacticMatch(source, target);
+        if (fullMatch && !isInJSFile(source)) {
+            Debug.assert(checker.getSymbolAtLocation(source) === checker.getSymbolAtLocation(target));
+        }
+        return fullMatch;
+    }
+
+    function isSyntacticMatch(source: Node, target: Node): boolean {

```

Comment:
This function is used in two mutually exclusive cases, so you'd be better off with it broken up:

In the first two calls, you know that the `isIdentifier` clause will never be true, because you just tested that at least `source` is a property access.
In the third call, you know the isPropertyAccessExpression clause will never be true, because you just left a while loop that exhausted that possibility.

I would inline all 3 cases and look for commonalities, though I'm not sure they'll appear.

Reply:
I removed the function and inlined those cases.